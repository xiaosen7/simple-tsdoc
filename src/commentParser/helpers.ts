import * as tsdoc from "@microsoft/tsdoc";
import { customTags } from "../customTags";

const allBlockTagNames = customTags
  .filter(({ syntaxKind }) => syntaxKind === tsdoc.TSDocTagSyntaxKind.BlockTag)
  .map(({ tagName }) => tagName)
  .concat(
    tsdoc.StandardTags.allDefinitions
      .filter(
        ({ syntaxKind }) => syntaxKind === tsdoc.TSDocTagSyntaxKind.BlockTag
      )
      .map(({ tagName }) => tagName)
  );

export const allBlockTagNamesWithTest = allBlockTagNames.map((name) => ({
  name,
  test: createRegByTagName(name),
}));

export const tsDocParser = createTSDocParser();

function createTSDocParser() {
  const customConfiguration: tsdoc.TSDocConfiguration =
    new tsdoc.TSDocConfiguration();

  // add  customTagDefinition
  customConfiguration.addTagDefinitions(
    customTags.map((x) => new tsdoc.TSDocTagDefinition(x))
  );

  return new tsdoc.TSDocParser(customConfiguration);
}

function createRegByTagName(tagName: string) {
  return new RegExp(tagName.slice(1), "i");
}
