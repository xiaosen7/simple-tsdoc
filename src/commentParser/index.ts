import * as types from "../types";
import * as helpers from "./helpers";
import * as shared from "../shared";

export function parse(content: string): types.Annotation {
  const parserContext = helpers.tsDocParser.parseString(content);
  const docComment = parserContext.docComment;

  const description = shared.formatDocNode(docComment.summarySection).trim();

  let deprecated: types.Annotation["deprecated"] = false;
  if (docComment.deprecatedBlock) {
    deprecated = shared
      .formatDocNode(docComment.deprecatedBlock.content)
      .trim();
  }

  const params: types.Annotation["params"] = [];
  for (const paramBlock of docComment.params.blocks) {
    params.push({
      name: paramBlock.parameterName,
      description: shared.formatDocNode(paramBlock.content).trim(),
    });
  }

  const annotationTags: types.Annotation["tags"] = [];
  [
    ...docComment.customBlocks,
    docComment.remarksBlock!,
    ...docComment.seeBlocks,
  ]
    .filter((x) => Boolean(x))
    .forEach((block) => {
      const found = helpers.allBlockTagNamesWithTest.find(({ test }) =>
        test.test(block.blockTag.tagName)
      );
      if (!found) {
        return;
      }

      const description = shared.formatDocNode(block.content).trim();
      annotationTags.push({
        name: found.name,
        description: description,
      });
    });

  const modifierTags = docComment.modifierTagSet.nodes.map((x) => x.tagName);

  let returns;
  if (docComment.returnsBlock) {
    returns = shared.formatDocNode(docComment.returnsBlock.content).trim();
  }

  return {
    description,
    params,
    tags: annotationTags,
    modifierTags,
    returns,
    deprecated,
  };
}
