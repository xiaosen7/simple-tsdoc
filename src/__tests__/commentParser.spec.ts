import * as commentParser from "../commentParser";
import * as testUtils from "./testUtils";

test("commentParser parse @constructor", () => {
  const content = testUtils.readAssetSync("comment.constructor.ts");
  const annotation = commentParser.parse(content);
  expect(annotation).toMatchSnapshot();
});

test("commentParser parse @returns", () => {
  const content = testUtils.readAssetSync("comment.returns.ts");
  const annotation = commentParser.parse(content);
  expect(annotation).toMatchSnapshot();
});

test("commentParser parse @example", () => {
  const content = testUtils.readAssetSync("comment.example.ts");
  const annotation = commentParser.parse(content);
  expect(annotation).toMatchSnapshot();
});

test("commentParser parse", () => {
  const content = testUtils.readAssetSync("comment.ts");
  const annotation = commentParser.parse(content);
  expect(annotation).toMatchSnapshot();
});
