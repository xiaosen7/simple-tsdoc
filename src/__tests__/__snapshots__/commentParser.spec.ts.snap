// Vitest Snapshot v1

exports[`commentParser parse @constructor 1`] = `
{
  "deprecated": false,
  "description": "",
  "modifierTags": [
    "@constructor",
  ],
  "namespace": "其他",
  "params": [],
  "returns": undefined,
  "tags": [],
}
`;

exports[`commentParser parse @example 1`] = `
{
  "deprecated": false,
  "description": "",
  "modifierTags": [],
  "namespace": "其他",
  "params": [],
  "returns": undefined,
  "tags": [
    {
      "description": "\`\`\`ts
console.log(\\"\\\\\`\\")
\`\`\`",
      "name": "@example",
    },
  ],
}
`;

exports[`commentParser parse @returns 1`] = `
{
  "deprecated": false,
  "description": "",
  "modifierTags": [],
  "namespace": "其他",
  "params": [],
  "returns": "返回描述",
  "tags": [],
}
`;

exports[`commentParser parse 1`] = `
{
  "deprecated": "use xxx instead",
  "description": "Returns the average of two numbers.
This is a second sentence.",
  "modifierTags": [
    "@beta",
  ],
  "namespace": "foobar",
  "params": [
    {
      "description": "The first input number",
      "name": "x",
    },
    {
      "description": "The second input number",
      "name": "y",
    },
  ],
  "returns": "The average of \`x\` and \`y\`",
  "tags": [
    {
      "description": "example 1

\`\`\`ts
console.log(\\"example\\");
\`\`\`

example 2

\`\`\`ts
console.log(\\"example2\\");
\`\`\`

example 3

\`\`\`vue
<template>
<button></button>
</template>
\`\`\`",
      "name": "@example",
    },
  ],
}
`;
